### PR TITLE
MM-3075 Add isValid method for openDialog request

### PR DESCRIPTION
**Pull request by Manimaran11**
#### Summary
MM-3075 Add isValid method for openDialog request

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/16199

#### Release Note
```release-note
Add validation that the payload for a open Interactive Dialog request is valid according to the rules at https://docs.mattermost.com/developer/interactive-dialogs.html
```

### CODE REVIEW

**Comment by Manimaran11** [comment]
@hanzei its WIP. I wanted to check with you if things are fine with this code or if any other changes are needed

File: model/integration_action.go [hunk 1]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+
+	if d.IntroductionText == "" {
+		return errors.New("invalid introduction text")
+	}
+
+	if len(d.Elements) > 5 {
+		return errors.New("invalid number of dialog elements")
+	}
+
+	if reflect.TypeOf(d.NotifyOnCancel).Kind() != reflect.Bool {
```
Comment (hanzei) [inline model/integration_action.go:569]: Why is this needed? `d.NotifyOnCancel` is typed as a `bool`.

File: model/integration_action.go [hunk 2]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		return errors.New("invalid type notify_on_cancel")
+	}
+
+	if len(d.Elements) != 0 {
+		elementMap := make(map[string]bool)
+
+		for i := 0; i < len(d.Elements); i++ {
+			if elementMap[d.Elements[i].DisplayName] == true {
+				return errors.New("duplicate dialog element name")
```
Comment (hanzei) [inline model/integration_action.go:578]: Nice way to check for that :+1:

File: model/integration_action.go [hunk 3]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+			}
+			elementMap[d.Elements[i].DisplayName] = true
+			dialogElementErr := d.Elements[i].isValid()
```
Comment (hanzei) [inline model/integration_action.go:581]: Nit pick: separate the two code sections

```suggestion
			elementMap[d.Elements[i].DisplayName] = true
			
			dialogElementErr := d.Elements[i].isValid()
```

File: model/integration_action.go [hunk 4]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+			if dialogElementErr != nil {
```
Comment (hanzei) [inline model/integration_action.go:582]: Nit pick: Why use a different variable name then `err`?

File: model/integration_action.go [hunk 5]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+				return dialogElementErr
+			}
+		}
+	}
+	return nil
+}
+
+func (d *DialogElement) isValid() error {
+	typeMap := map[string]bool{
```
Comment (hanzei) [inline model/integration_action.go:591]: Maybe `textSubTypeMap`?

File: model/integration_action.go [hunk 6]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		"text":   true,
+		"email":  true,
+		"number": true,
+		"tel":    true,
+		"url":    true,
+	}
+
+	switch d.Type {
+
+	case "email":
```
Comment (hanzei) [inline model/integration_action.go:601]: Actually, the `type` is `text`, not `email`.

File: model/integration_action.go [hunk 7]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+	typeMap := map[string]bool{
+		"text":   true,
+		"email":  true,
+		"number": true,
+		"tel":    true,
+		"url":    true,
+	}
+
+	switch d.Type {
+
```
Comment (hanzei) [inline model/integration_action.go:600]: Nit pick: Remove this line

File: model/integration_action.go [hunk 8]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+	}
+	return nil
+}
+
+func (d *DialogElement) isValid() error {
+	typeMap := map[string]bool{
+		"text":   true,
+		"email":  true,
+		"number": true,
+		"tel":    true,
```
Comment (hanzei) [inline model/integration_action.go:595]: `password` seems to be missing

File: model/integration_action.go [hunk 9]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		"url":    true,
+	}
+
+	switch d.Type {
+
+	case "email":
+		if d.MaxLength > 150 {
```
Comment (hanzei) [inline model/integration_action.go:602]: Well, the documentation doesn't state the element as invalid if `MaxLength` is longer then 150.

File: model/integration_action.go [hunk 10]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		if d.MaxLength > 3000 {
+			return errors.New("invalid maxlength")
+		}
+		if d.Placeholder != "" && len(d.Placeholder) > 3000 {
+			return errors.New("invalid placeholder")
+		}
+		if d.Default != "" && len(d.Placeholder) > 3000 {
+			return errors.New("invalid default string")
+		}
+		if d.HelpText != "" && len(d.HelpText) > 3000 {
```
Comment (hanzei) [inline model/integration_action.go:628]: I think `d.HelpText` should always contains a maximum of 150 characters.

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for digging deep into Interactive dialogs! A lot of comments about style and some semantic one below.

File: model/integration_action.go (hunk 9 above)
Comment (Manimaran11) [reply model/integration_action.go:602]: for type text, if the max length is longer than 150, the suggestion is to use textarea. Should I remove this ??

File: model/integration_action.go (hunk 2 above)
Comment (Manimaran11) [reply model/integration_action.go:578]: Thanks :)

File: model/integration_action.go [hunk 11]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+func (d *Dialog) isValid() error {
+
+	if d.Title == "" || len(d.Title) > 24 {
+		return errors.New("invalid dialog title")
+	}
+
+	if d.IconURL != "" && !IsValidHttpUrl(d.IconURL) {
+		return errors.New("invalid icon url")
+	}
+
```
[the remark this answers is not in the document]
Comment (Manimaran11) [reply model/integration_action.go:560]: the dialog doesnt have URL .

File: model/integration_action.go [hunk 12]
```diff
@@ -530,3 +532,142 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
 	json.NewDecoder(data).Decode(&o)
 	return o
 }
+
+func (r *OpenDialogRequest) isValid() error {
```
[the remark this answers is not in the document]
Comment (Manimaran11) [reply model/integration_action.go:536]: how to get the App from the isValid method

File: model/integration_action.go (hunk 9 above)
Comment (hanzei) [reply model/integration_action.go:602]: Yes, for the 'email` has this check has to be removed.

File: model/integration_action.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply model/integration_action.go:536]: You need to export `isValid` first by naming it to `IsValid`. That allows the `App` to import it.

**Comment by jasonblais** [comment]
Thank you @Manimaran11 for the updates! It looks like some of the tests are failing, are you open to help take a look to resolve?

**Comment by Manimaran11** [comment]
Definitely will look into it
Thanks
Manimaran

File: app/integration_action.go [hunk 13]
```diff
@@ -573,6 +573,10 @@ func (a *App) OpenInteractiveDialog(request model.OpenDialogRequest) *model.AppE
 		return err
 	}
 
+	if dialogErr := request.IsValid(); dialogErr != nil {
+		return dialogErr
+	}
+
 	request.TriggerId = clientTriggerId
```
Comment (Manimaran11) [inline app/integration_action.go:580]: @hanzei adding the isValid method in openDialog is leading to error, as the dialogErr returns an error and the openInteractiveDialog returns model.AppError. Any suggestions??

**Comment by jfrerich** [comment]
@hanzei, gentle ping for a response to the Author

File: app/integration_action.go (hunk 13 above)
Comment (hanzei) [reply app/integration_action.go:580]: Using `model.NewAppError` you can wrap the returned `error` into an `model.AppError` similar how it's done in https://github.com/mattermost/mattermost-server/pull/16358/files/796c8226667c30d0d1becae347a7297f87d78ca2#diff-47384b35a1503accad7764e31e11043e2a8aaa220a7fa8b03009f2704346512cR597-R599. That also means that you have to add a new translation string to ``i18n/en.json`.

**Review by hanzei — DISMISSED** [review]
Blocking until tests failures are resolved. Let me know if you need any further help!

File: app/integration_action.go (hunk 13 above)
Comment (Manimaran11) [reply app/integration_action.go:580]: this points to the current PR. Can you please direct me to the correct url

**Comment by emilyacook** [comment]
> Blocking until tests failures are resolved. Let me know if you need any further help!

@Manimaran11 Hello! Can you please let me know what the status is on this PR so I can update it as per the [inactive contribution guidelines](https://developers.mattermost.com/contribute/getting-started/inactive-contributions/)?

**Comment by Manimaran11** [comment]
@emilyacook, really sorry for the delay. Will finish it by this week.

**Comment by emilyacook** [comment]
@Manimaran11 No worries! Don't feel pressured, I just need to update the status if it's going to take some time 🙂

File: app/integration_action.go (hunk 13 above)
Comment (hanzei) [reply app/integration_action.go:580]: The line I wanted to point out is: https://github.com/mattermost/mattermost-server/blob/796c8226667c30d0d1becae347a7297f87d78ca2/app/integration_action.go#L596-L599

File: model/integration_action.go [hunk 14]
```diff
@@ -530,3 +532,125 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		}
+		if d.Default != "" && len(d.Placeholder) > 3000 {
+			return errors.New("invalid default string")
+		}
+		if _, ok := textSubTypeMap[d.Type]; !ok {
+			return errors.New("invalid subtype")
+		}
+
+	case "checkbox":
+		if len(d.DataSource) == 0 {
```
Comment (hanzei) [inline model/integration_action.go:622]: This will fix CI
```suggestion
		if d.DataSource == "" {
```

**Comment by Manimaran11** [comment]
Can someone else please look into the PR
Thanks
Manimaran

File: model/integration_action.go [hunk 15]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+
+	switch d.Type {
+	case "text":
+		if len(d.Default) > 150 {
+			return errors.New("invalid default text")
+		}
+		if len(d.Placeholder) > 150 {
+			return errors.New("invalid placeholder")
+		}
+		if len(d.Placeholder) > 150 {
```
Comment (larkox) [inline model/integration_action.go:608]: This seems repeated here and on the textarea. Not sure if it was looking for a different field.

File: model/integration_action.go [hunk 16]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+			return errors.New("invalid placeholder")
+		}
+		if len(d.Placeholder) > 3000 {
+			return errors.New("invalid default string")
+		}
+		if _, ok := textSubTypeMap[d.Type]; !ok {
+			return errors.New("invalid subtype")
+		}
+
+	case "select":
```
Comment (larkox) [inline model/integration_action.go:629]: Should we check there is at least one option?

File: model/integration_action.go [hunk 17]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
 	json.NewDecoder(data).Decode(&o)
 	return o
 }
+
+func (r *OpenDialogRequest) IsValid() error {
```
Comment (larkox) [inline model/integration_action.go:536]: Everything in this file seems to agree with the documentation, but I wonder if we are being so strict in the code, and if adding these restrictions may break working integrations.

**Review by larkox — CHANGES_REQUESTED** [review]

File: model/integration_action.go [hunk 18]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+	}
+
+	if len(d.Elements) > 5 {
+		return errors.New("The maximum number of dialog elements is 5")
+	}
+
+	if len(d.Elements) != 0 {
+		elementMap := make(map[string]bool)
+
+		for i := 0; i < len(d.Elements); i++ {
```
Comment (mickmister) [inline model/integration_action.go:564]: Probably best to use a `range` loop here for better readability inside of the loop

```suggestion
		for i, element := range d.Elements {
```

File: model/integration_action.go [hunk 19]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		"text":     true,
+		"email":    true,
+		"number":   true,
+		"tel":      true,
+		"url":      true,
+		"password": true,
+	}
+
+	if d.MinLength < 0 {
+		return errors.New("invalid minLength")
```
Comment (mickmister) [inline model/integration_action.go:591]: It's probably best to use the JSON keys in these errors, as these will be sent back via HTTP error. The caller could be written in any language, so JSON seems to be the best representation to refer to in the response.

```suggestion
		return errors.New("min_length cannot be a negative number")
```

File: model/integration_action.go [hunk 20]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		return errors.New("invalid maxlength")
+	}
+	if d.MinLength > d.MaxLength {
+		return errors.New("minLength should not be greater then maxlength")
+	}
+
+	switch d.Type {
+	case "text":
+		if len(d.Default) > 150 {
+			return errors.New("invalid default text")
```
Comment (mickmister) [inline model/integration_action.go:603]: We should be descriptive about why the text is invalid. Please go through other error strings in the PR as well and apply the same principle where the error message can be a bit vague

```suggestion
			return errors.Errorf("default text cannot be longer than %d characters", MAX_DIALOG_DEFAULT_TEXT_LENGTH)
```

File: model/integration_action.go [hunk 21]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		}
+		if len(d.Placeholder) > 150 {
+			return errors.New("invalid placeholder")
+		}
+		if len(d.Placeholder) > 150 {
+			return errors.New("invalid default string")
+		}
+		if _, ok := textSubTypeMap[d.Type]; !ok {
+			return errors.New("invalid subtype")
```
Comment (mickmister) [inline model/integration_action.go:612]:
```suggestion
			return errors.Errorf("invalid subtype '%s'", d.Type)
```

File: model/integration_action.go [hunk 22]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+		}
+
+	case "textarea":
+		if len(d.Default) > 3000 {
+			return errors.New("invalid default text")
```
Comment (mickmister) [inline model/integration_action.go:617]: Should we also check that `d.Default` is within the bounds of `MinLength` and `MaxLength`?

File: model/integration_action.go (hunk 16 above)
Comment (mickmister) [reply model/integration_action.go:629]: Agreed

File: model/integration_action.go [hunk 23]
```diff
@@ -530,3 +532,136 @@ func DoPostActionRequestFromJson(data io.Reader) *DoPostActionRequest {
...
+	case "select":
+		if len(d.Default) > 3000 {
+			return errors.New("invalid default string")
+		}
+		if len(d.Placeholder) > 3000 {
+			return errors.New("invalid placeholder")
+		}
+
+	case "checkbox":
+		if len(d.Default) > 150 {
```
Comment (mickmister) [inline model/integration_action.go:638]: Lots of magic numbers going on here. Maybe make a function to deduplicate the hardcoded numbers and logic?

**Review by mickmister — CHANGES_REQUESTED** [review]
Great work @Manimaran11 :+1: I have some requests regarding checking unique names and wording of errors

**Comment by hanzei** [comment]
Closing as the PR has become orphaned.
